dali/python/backend_impl.cc

```diff
@@ -296,8 +296,15 @@ void FillTensorFromCudaArray(const py::object object, TensorType *batch, int dev
 
   // Keep a copy of the input object ref in the deleter, so its refcount is increased
   // while this shared_ptr is alive (and the data should be kept alive)
-  batch->ShareData(shared_ptr<void>(ptr, [obj_ref = object](void *) {}),
-                   bytes, false, typed_shape, type.id(), device_id);
+  batch->ShareData(shared_ptr<void>(ptr, [obj_ref = object](void *) mutable {  // NOLINT
+    py::gil_scoped_acquire aqr;
+    {
+      auto tmp = std::move(obj_ref);
+      (void)tmp;
+    }
```

Comment:
This lambda is quite unclear - e.g. why the `(void)tmp;`. Could you add a description what happens there and why?

Reply:
Will do.

Reply:
This "trick" stores the object whose lifetime we want to prolong in the lambda closure. When the shared pointer is destroyed, the deleter (the lambda) is invoked. In the lambda, we acquire GIL and destroy the contents of the closure while GIL is held. I've described it in the comments, too.